soc/usbexecd.c

```diff
@@ -0,0 +1,1029 @@
+
+            // Create a writable pipebuf and return the readable pipe for the child process
+            int ctrl_fd   = pipebuf_out_init (&p->ctrl,   id, 0);
+            int stdin_fd  = pipebuf_out_init (&p->stdin,  id, 1);
+
+            // Create a readable pipebuf and return the writable pipe for the child process
+            int stdout_fd = pipebuf_in_init(&p->stdout, id, 2);
+            int stderr_fd = pipebuf_in_init(&p->stderr, id, 3);
+
+            int pid;
+            // Fork the child process, if this is the child process
+            if ((pid = fork()) == 0) {
+
+                // For every process in the parent
+                for (int iterator = 0; iterator < N_PROC; iterator++) {
+                    // Get a reference to the correspondong process
+                    procinfo_t* parent_process = processes[iterator];
+
+                    // If this process exists
+                    if (parent_process) {
+                        // Close all of the file descriptors for this child
+                        if ((&parent_process->ctrl)->fd != -1) close((&parent_process->ctrl)->fd);
+                        if ((&parent_process->stdin)->fd != -1) close((&parent_process->stdin)->fd);
+                        if ((&parent_process->stdout)->fd != -1) close((&parent_process->stdout)->fd);
+                        if ((&parent_process->stderr)->fd != -1) close((&parent_process->stderr)->fd);
+                    }
+                }
+
+                // Close socket and fds used by parent
+                close(sock);
+                close(epfd);
+                close(sigfd);
+
+                // Get started on its own task
+                child(ctrl_fd, stdin_fd, stdout_fd, stderr_fd);
+            // If there was an error
+            } else if (pid < 0) {
+                // Notify the user
+                fatal("Error in fork: %s", strerror(errno));
+            // If this is the parent process 
+            } else {
+                // Close the child ends of the pipes
+                close(ctrl_fd);
+                close(stdin_fd);
+                close(stdout_fd);
+                close(stderr_fd);
+                // Set the field for the child process pid
+                p->pid = pid;
+            }
+
+            break;
+
+        case CMD_CLOSE:
+            debug("CMD: Close a process with %d", id);
+            // If the process isn't killed yet
+            if (p->pid) {
+                // Kill it now
+                kill(p->pid, SIGKILL);
+                // Wait for it to be finished
+                waitpid(p->pid, NULL, 0);
+            }
+
+            // Close out all of the pipe buffers if they haven't been closed already
+            if ((&p->ctrl)->fd != -1) pipebuf_out_close(&p->ctrl);
+            if ((&p->stdin)->fd != -1) pipebuf_out_close(&p->stdin);
+            if ((&p->stdout)->fd != -1) pipebuf_in_close(&p->stdout);
+            if ((&p->stderr)->fd != -1) pipebuf_in_close(&p->stderr);
```

Comment:
I'll go ahead and add a `force` flag to the function definition to differentiate between a casual close and a must close.